lib/system/assertions.nim

```diff
@@ -113,3 +113,42 @@ template doAssertRaises*(exception: typedesc, code: untyped) =
     except: raisedForeign()
   if wrong:
     raiseAssert(begin & " nothing was raised" & msgEnd)
+
+# ------------- unreachable -------------
+
+func unreachableImpl(str: string, loc: InstantiationInfo) {.
+    noinline, noreturn.} =
+  var msg: string
+  msg.toLocation(loc.filename, loc.line, loc.column + 1)
+  msg.add:
+    if str.len > 0: " unreachable: "
+    else:           " unreachable"
+  msg.add str
+  raiseAssert(msg)
+
+func unreachableImpl(e: enum, loc: InstantiationInfo) {.noinline, noreturn.} =
+  ## A bit more efficient than ``unreachable($e)``, as the stringification
+  ## code is not part of the callsite (less I-cache pressure).
+  unreachableImpl($e, loc)
+
+template unreachable*() =
+  ## Marks a point in the program as being unreachable. At run-time, behaves
+  ## the same as:
+  ##
+  ## .. code-block:: nim
+  ##
+  ##   doAssert false, "unreachable"
+  ##
+  ## but the intention is clearer and the call marks the execution path as not
+  ## returning anything.
```

Comment:
I think it's worth highlighting that this is a good way to handle impossible cases in a conditional expression like case statement.

The example I wrote here is not very good, but the idea here is that some examples would be helpful.

```suggestion
  ## but the intention is clearer and the call marks the execution path as not
  ## returning anything and can be used within conditional expressions.
  runnableExamples:
    type Direction = enum Up, Down, Left, Right
    proc oppositeX(dir: Direction): Direction =
      case dir
      of Left: Right
      of Right: Left
      else: unreachable()
```

Reply:
I'll add an example.

Reply:
This example might work better.

```nim
runnableExamples:
  let arr = [1, 2, 3, 4, 5]
  for i in arr.items:
    if i mod 2 == 0:
      case i
      of 2: echo "two"
      of 4: echo "four"
      else: unreachable()
```

Reply:
I've used the `mod` example with two small adjustments:
* `1..5` instead of an extra array
* a case expression instead of a case statement, as the case statement would also compile with `unreachable` replaced by `discard`